pootle/static/js/editor/UnitList.js

```diff
@@ -0,0 +1,294 @@
+  }
+
+  gotoPosition(pos) {
+    if (this.uids.length === 0) {
+      return;
+    }
+
+    let idx = pos - this.begin - 1;
+
+    if (idx < 0) {
+      idx = 0;
+    }
+    if (idx > this.uids.length - 1) {
+      idx = this.uids.length - 1;
+    }
+
+    this.goto(this.uids[idx]);
+  }
+
+  onFirstUnit() {
+    return this.idx <= 0; // -1 or 0
+  }
+
+  onLastUnit() {
+    return this.idx === this.uids.length - 1;
+  }
+
+  firstUnitInVicinity(vicinity) {
+    return Math.max(0, this.idx - vicinity);
+  }
+
+  lastUnitInVicinity(vicinity) {
+    return Math.min(this.uids.length - 1, this.idx + vicinity);
+  }
+
+  visibleUnitsBefore() {
+    // TODO: instead of just reporting unis to render,
+    // can we use a callback whenever the visible units change?
+    const start = this.firstUnitInVicinity(this.visibleRowsBefore);
+    const end = this.idx - 1;
+    return this.uids.slice(start, end + 1);
+  }
+
+  visibleUnitsAfter() {
+    // TODO: instead of just reporting unis to render,
+    // can we use a callback whenever the visible units change?
+    const start = this.idx + 1;
+    const end = this.lastUnitInVicinity(this.visibleRowsAfter);
+    return this.uids.slice(start, end + 1);
+  }
+
+  prefetchUnits() {
+    const start = this.firstUnitInVicinity(this.visibleRowsBefore + this.prefetchRows);
+    const end = this.lastUnitInVicinity(this.visibleRowsAfter + this.prefetchRows);
+    const uids = this.uids.slice(start, end + 1);
+    const originalUid = this.uid;
+
+    const filteredUids = [];
+
+    for (let i = 0; i < uids.length; i++) {
+      const uid = uids[i];
+      if (!this.units[uid]) {
+        filteredUids.push(uid);
+      }
+    }
+
+    if (!filteredUids.length) {
+      return; // nothing to do
+    }
+
+    UnitAPI.fetchViewRows({ uids: filteredUids }).then(
+      (data) => {
+        let hasNewData;
+
+        this.ready = true;
+
+        if (!data.rows) {
+          return;
+        }
+
+        Object.keys(data.rows).forEach((uid) => {
+          if (this.units[uid]) {
+            return; // unit was already loaded
+          }
+          hasNewData = true;
+          this.units[uid] = new Unit(data.rows[uid]);
+        });
+
+        if (hasNewData) {
+          // TODO: fire this only if the rows in the viewport have changed
+          this.handleViewRowsChange(originalUid);
+        }
+      },
+      this.error
+    );
+  }
+
+  cleanupUnusedUnits() {
+    const start1 = 0;
+    const end1 = this.firstUnitInVicinity(
+      this.visibleRowsBefore + this.prefetchRows
+    ) - 1;
+    this.cleanupUnitsInRange(start1, end1);
+
+    const start2 = this.lastUnitInVicinity(
+      this.visibleRowsAfter + this.prefetchRows
+    ) + 1;
+    const end2 = this.length - 1;
+    this.cleanupUnitsInRange(start2, end2);
+  }
+
+  cleanupUnitsInRange(start, end) {
+    for (let i = start; i <= end; i++) {
+      const uid = this.uids[i];
+      if (this.units[uid]) {
+        delete this.units[uid];
```

Comment:
Here we're cleaning up the unit that are not visible and are not used, so by design no outside code should keep any references to units. The idea behind creating UnitsList class was to encapsulate all the logic of working with units so that outside code wouldn't deal will keeping any references.